Sources/PackageModel/DependencyMapper.swift

```diff
@@ -29,143 +28,220 @@ public struct DefaultDependencyMapper: DependencyMapper {
-    ///
-    /// This is *not* a generic URI scheme parser!
-    private static func parseScheme(_ location: String) -> String? {
-        func prefixOfSplitBy(_ delimiter: String) -> String? {
-            let (head, tail) = location.spm_split(around: delimiter)
-            if tail == nil {
-                //not found
-                return nil
-            } else {
-                //found, return head
-                //lowercase the "scheme", as specified by the URI RFC (just in case)
-                return head.lowercased()
+    private func normalizeDependencyLocation(
+        dependency: MappablePackageDependency,
+        parentPackagePath: AbsolutePath,
+        fileSystem: FileSystem
+    ) throws -> String? {
+        switch dependency.kind {
+        // nothing to normalize
+        case .registry:
+            return .none
+        // location may be a relative path so we need to normalize it
+        case .fileSystem, .sourceControl:
+            let dependencyLocation = dependency.locationString
+            switch parseScheme(dependencyLocation) {
+            // if the location has no scheme, we treat it as a path (either absolute or relative).
+            case .none:
+                // if the dependency URL starts with '~/', try to expand it.
+                if dependencyLocation.hasPrefix("~/") {
+                    return try AbsolutePath(validating: String(dependencyLocation.dropFirst(2)), relativeTo: fileSystem.homeDirectory).pathString
+                }
+
+                // check if already absolute path
+                if let path = try? AbsolutePath(validating: dependencyLocation) {
+                    return path.pathString
+                }
+
+                // otherwise treat as relative path to the parent package
+                return try AbsolutePath(validating: dependencyLocation, relativeTo: parentPackagePath).pathString
+            // SwiftPM can't handle file locations with file:// scheme so we need to
+            // strip that. We need to design a Location data structure for SwiftPM.
+            case .some("file"):
+                let location = String(dependencyLocation.dropFirst(Self.filePrefix.count))
+                let hostnameComponent = location.prefix(while: { $0 != "/" })
+                guard hostnameComponent.isEmpty else {
+                  if hostnameComponent == ".." {
+                      throw DependencyMappingError.invalidFileURL("file:// URLs cannot be relative, did you mean to use '.package(path:)'?")
+                  }
+                  throw DependencyMappingError.invalidFileURL("file:// URLs with hostnames are not supported, are you missing a '/'?")
+                }
+                return try AbsolutePath(validating: location).pathString
+            // if the location has a scheme, assume a URL and nothing to normalize
+            case .some(_):
+                return .none
             }
         }
+    }
+}
+
+// trivial representation for mapping
```

Comment:
Can this be converted to a doc comment with more details describing what exactly is being mapped?

Reply:
this is an internal utility for applying mirrors, tbh I think this is pretty clear from the name and code what this does so I dont think adding docs would add value

Reply:
Having a quick look at the code I didn't get that it has a connection to mirrors. I hope this could mention that it's related to mirrors at least. While it's internal, there's still huge value in such documentation, as it makes it easier to review and easier to circle back to it later when needed.

Reply:
added comment to the main utility